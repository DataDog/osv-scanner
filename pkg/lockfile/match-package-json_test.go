package lockfile_test

import (
	"github.com/google/osv-scanner/internal/testutility"
	"io/fs"
	"os"
	"path/filepath"
	"testing"

	"github.com/google/osv-scanner/pkg/lockfile"
	"github.com/stretchr/testify/assert"
)

var packageJSONMatcher = lockfile.PackageJSONMatcher{}

func TestPackageJSONMatcher_GetSourceFile_FileDoesNotExist(t *testing.T) {
	t.Parallel()

	lockFile, err := lockfile.OpenLocalDepFile("fixtures/package-json/does-not-exist/npm-v1.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := packageJSONMatcher.GetSourceFile(lockFile)
	expectErrIs(t, err, fs.ErrNotExist)
	assert.Equal(t, "", sourceFile.Path())
}

func TestPackageJSONMatcher_GetSourceFile(t *testing.T) {
	t.Parallel()
	dir, err := os.Getwd()
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	basePath := "fixtures/package-json/one-package/"
	sourcefilePath := filepath.FromSlash(filepath.Join(dir, basePath+"package.json"))

	lockFile, err := lockfile.OpenLocalDepFile(basePath + "npm-v1.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	sourceFile, err := packageJSONMatcher.GetSourceFile(lockFile)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	assert.Equal(t, sourcefilePath, sourceFile.Path())
}

func TestPackageJSONMatcher_Match_OnePackage(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/package-json/one-package/package.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:           "lodash",
			TargetVersions: []string{"^4.0.0"},
		},
	}
	err = packageJSONMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	testutility.NewSnapshot().MatchText(t, testutility.NormalizeJSON(t, packages))
}

func TestPackageJSONMatcher_Match_TransitiveDependencies(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/package-json/transitive/package.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:           "commander",
			TargetVersions: []string{"~2.0.0"},
		},
		{
			Name:           "debug",
			TargetVersions: []string{"^0.7", "~0.7.2"},
		},
		{
			Name:           "jear",
			TargetVersions: []string{"^0.1.4"},
		},
		{
			Name:           "shelljs",
			TargetVersions: []string{"~0.1.4"},
		},
		{
			Name:           "velocityjs",
			TargetVersions: []string{"~0.3.15"},
		},
	}
	err = packageJSONMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	testutility.NewSnapshot().MatchText(t, testutility.NormalizeJSON(t, packages))
}

func TestPackageJSONMatcher_Match_NameConflict(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/package-json/name-conflict/package.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:           "aws-sdk-client-mock",
			TargetVersions: []string{"^2.1.1"},
		},
		{
			Name:           "aws-sdk-client-mock-jest",
			TargetVersions: []string{"^2.1.1"},
		},
	}
	err = packageJSONMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	testutility.NewSnapshot().MatchText(t, testutility.NormalizeJSON(t, packages))
}

func TestPackageJSONMatcher_Match_Resolutions(t *testing.T) {
	t.Parallel()

	sourceFile, err := lockfile.OpenLocalDepFile("fixtures/package-json/resolutions/package.json")
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	packages := []lockfile.PackageDetails{
		{
			Name:           "fast-xml-parser",
			Version:        "4.2.5",
			TargetVersions: []string{"4.2.5"},
		},
		{
			Name:           "fast-xml-parser",
			Version:        "4.4.0",
			TargetVersions: []string{"^4.2.5"},
		},
		{
			Name:           "@aws-sdk/core",
			Version:        "3.535.0",
			TargetVersions: []string{"^3.535.0"},
		},
	}
	err = packageJSONMatcher.Match(sourceFile, packages)
	if err != nil {
		t.Errorf("Got unexpected error: %v", err)
	}

	testutility.NewSnapshot().MatchText(t, testutility.NormalizeJSON(t, packages))
}
