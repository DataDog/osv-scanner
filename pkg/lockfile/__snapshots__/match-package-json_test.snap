
[TestPackageJSONMatcher_Match_NameConflict - 1]
[
  {
    "name": "aws-sdk-client-mock",
    "version": "",
    "targetVersions": [
      "^2.1.1"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 5,
        "end": 36
      },
      "file_name": "<rootdir>/fixtures/package-json/name-conflict/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 29,
        "end": 35
      },
      "file_name": "<rootdir>/fixtures/package-json/name-conflict/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 6,
        "end": 25
      },
      "file_name": "<rootdir>/fixtures/package-json/name-conflict/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "aws-sdk-client-mock-jest",
    "version": "",
    "targetVersions": [
      "^2.1.1"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 5,
        "end": 41
      },
      "file_name": "<rootdir>/fixtures/package-json/name-conflict/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 34,
        "end": 40
      },
      "file_name": "<rootdir>/fixtures/package-json/name-conflict/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 6,
        "end": 30
      },
      "file_name": "<rootdir>/fixtures/package-json/name-conflict/package.json"
    },
    "isDirect": true
  }
]
---

[TestPackageJSONMatcher_Match_OnePackage - 1]
[
  {
    "name": "lodash",
    "version": "",
    "targetVersions": [
      "^4.0.0"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 5,
        "end": 23
      },
      "file_name": "<rootdir>/fixtures/package-json/one-package/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 16,
        "end": 22
      },
      "file_name": "<rootdir>/fixtures/package-json/one-package/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 6,
        "end": 12
      },
      "file_name": "<rootdir>/fixtures/package-json/one-package/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  }
]
---

[TestPackageJSONMatcher_Match_Resolutions - 1]
[
  {
    "name": "fast-xml-parser",
    "version": "4.2.5",
    "targetVersions": [
      "4.2.5"
    ],
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "fast-xml-parser",
    "version": "4.4.0",
    "targetVersions": [
      "^4.2.5"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 5,
        "end": 32
      },
      "file_name": "<rootdir>/fixtures/package-json/resolutions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 25,
        "end": 31
      },
      "file_name": "<rootdir>/fixtures/package-json/resolutions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 6,
        "end": 21
      },
      "file_name": "<rootdir>/fixtures/package-json/resolutions/package.json"
    },
    "isDirect": true
  },
  {
    "name": "@aws-sdk/core",
    "version": "3.535.0",
    "targetVersions": [
      "^3.535.0"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 8,
        "end": 8
      },
      "column": {
        "start": 5,
        "end": 32
      },
      "file_name": "<rootdir>/fixtures/package-json/resolutions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 8,
        "end": 8
      },
      "column": {
        "start": 23,
        "end": 31
      },
      "file_name": "<rootdir>/fixtures/package-json/resolutions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 8,
        "end": 8
      },
      "column": {
        "start": 6,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/resolutions/package.json"
    },
    "isDirect": true
  }
]
---

[TestPackageJSONMatcher_Match_Target_Version - 1]
[
  {
    "name": "foo",
    "version": "1.5.3",
    "targetVersions": [
      "1.0.0 - 2.9999.9999"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 3,
        "end": 3
      },
      "column": {
        "start": 5,
        "end": 33
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 3,
        "end": 3
      },
      "column": {
        "start": 13,
        "end": 32
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 3,
        "end": 3
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "bar",
    "version": "1.5.3",
    "targetVersions": [
      "/u003e=1.0.2 /u003c2.1.2"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 5,
        "end": 28
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 13,
        "end": 27
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "baz",
    "version": "1.5.3",
    "targetVersions": [
      "/u003e1.0.2 /u003c=2.3.4"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 5,
        "end": 28
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 13,
        "end": 27
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "boo",
    "version": "1.5.3",
    "targetVersions": [
      "1.5.3"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 15,
        "end": 15
      },
      "column": {
        "start": 5,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 15,
        "end": 15
      },
      "column": {
        "start": 13,
        "end": 18
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 15,
        "end": 15
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "qux",
    "version": "1.5.3",
    "targetVersions": [
      "/u003c1.0.0 || /u003e=2.3.1 /u003c2.4.5 || /u003e=2.5.2 /u003c3.0.0"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 6,
        "end": 6
      },
      "column": {
        "start": 5,
        "end": 56
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 6,
        "end": 6
      },
      "column": {
        "start": 13,
        "end": 55
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 6,
        "end": 6
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "asd",
    "version": "1.5.3",
    "targetVersions": [
      "http://asdf.com/asdf.tar.gz"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 5,
        "end": 41
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 13,
        "end": 40
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 7,
        "end": 7
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "til",
    "version": "1.5.3",
    "targetVersions": [
      "~1.5"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 8,
        "end": 8
      },
      "column": {
        "start": 5,
        "end": 18
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 8,
        "end": 8
      },
      "column": {
        "start": 13,
        "end": 17
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 8,
        "end": 8
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "elf",
    "version": "1.5.3",
    "targetVersions": [
      "~1.5.3"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 9,
        "end": 9
      },
      "column": {
        "start": 5,
        "end": 20
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 9,
        "end": 9
      },
      "column": {
        "start": 13,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 9,
        "end": 9
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "two",
    "version": "1.5.3",
    "targetVersions": [
      "1.x"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 10,
        "end": 10
      },
      "column": {
        "start": 5,
        "end": 17
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 10,
        "end": 10
      },
      "column": {
        "start": 13,
        "end": 16
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 10,
        "end": 10
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "thr",
    "version": "1.5.3",
    "targetVersions": [
      "1.5.x"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 11,
        "end": 11
      },
      "column": {
        "start": 5,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 11,
        "end": 11
      },
      "column": {
        "start": 13,
        "end": 18
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 11,
        "end": 11
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "lat",
    "version": "1.5.3",
    "targetVersions": [
      "latest"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 12,
        "end": 12
      },
      "column": {
        "start": 5,
        "end": 20
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 12,
        "end": 12
      },
      "column": {
        "start": 13,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 12,
        "end": 12
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "dyl",
    "version": "1.5.3",
    "targetVersions": [
      "file:../dyl"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 13,
        "end": 13
      },
      "column": {
        "start": 5,
        "end": 25
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 13,
        "end": 13
      },
      "column": {
        "start": 13,
        "end": 24
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 13,
        "end": 13
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "kpg",
    "version": "1.5.3",
    "targetVersions": [
      "npm:pkg@1.5.0"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 14,
        "end": 14
      },
      "column": {
        "start": 5,
        "end": 27
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 14,
        "end": 14
      },
      "column": {
        "start": 13,
        "end": 26
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 14,
        "end": 14
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "abc",
    "version": "1.5.3",
    "targetVersions": [
      "1"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 17,
        "end": 17
      },
      "column": {
        "start": 5,
        "end": 15
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 17,
        "end": 17
      },
      "column": {
        "start": 13,
        "end": 14
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 17,
        "end": 17
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "cde",
    "version": "1.5.3",
    "targetVersions": [
      "/u003e1.0.2"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 16,
        "end": 16
      },
      "column": {
        "start": 5,
        "end": 20
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 16,
        "end": 16
      },
      "column": {
        "start": 13,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 16,
        "end": 16
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "dd",
    "version": "0.0.0-use.local",
    "targetVersions": [
      "javascript/datadog"
    ],
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "dd2",
    "version": "0.0.0-use.local",
    "targetVersions": [
      "javascript/datadog"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 19,
        "end": 19
      },
      "column": {
        "start": 5,
        "end": 32
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 19,
        "end": 19
      },
      "column": {
        "start": 13,
        "end": 31
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 19,
        "end": 19
      },
      "column": {
        "start": 6,
        "end": 9
      },
      "file_name": "<rootdir>/fixtures/package-json/multiple-versions/package.json"
    },
    "packageManager": "NPM",
    "isDirect": true
  }
]
---

[TestPackageJSONMatcher_Match_TransitiveDependencies - 1]
[
  {
    "name": "commander",
    "version": "",
    "targetVersions": [
      "~2.0.0"
    ],
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "packageManager": "NPM",
    "isDirect": true
  },
  {
    "name": "debug",
    "version": "",
    "targetVersions": [
      "^0.7",
      "~0.7.2"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 5,
        "end": 20
      },
      "file_name": "<rootdir>/fixtures/package-json/transitive/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 15,
        "end": 19
      },
      "file_name": "<rootdir>/fixtures/package-json/transitive/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 5,
        "end": 5
      },
      "column": {
        "start": 6,
        "end": 11
      },
      "file_name": "<rootdir>/fixtures/package-json/transitive/package.json"
    },
    "isDirect": true
  },
  {
    "name": "jear",
    "version": "",
    "targetVersions": [
      "^0.1.4"
    ],
    "depGroups": [
      "prod"
    ],
    "blockLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 5,
        "end": 21
      },
      "file_name": "<rootdir>/fixtures/package-json/transitive/package.json"
    },
    "versionLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 14,
        "end": 20
      },
      "file_name": "<rootdir>/fixtures/package-json/transitive/package.json"
    },
    "nameLocation": {
      "line": {
        "start": 4,
        "end": 4
      },
      "column": {
        "start": 6,
        "end": 10
      },
      "file_name": "<rootdir>/fixtures/package-json/transitive/package.json"
    },
    "isDirect": true
  },
  {
    "name": "shelljs",
    "version": "",
    "targetVersions": [
      "~0.1.4"
    ],
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "isDirect": true
  },
  {
    "name": "velocityjs",
    "version": "",
    "targetVersions": [
      "~0.3.15"
    ],
    "blockLocation": {
      "line": {
        "start": 0,
        "end": 0
      },
      "column": {
        "start": 0,
        "end": 0
      },
      "file_name": ""
    },
    "isDirect": true
  }
]
---
